### ISSUE TITLE
Mouse events randomly fail to register

### ISSUE THREAD

**Issue by cyardimci** [pixijs/pixi.js#3616]
Tested on versions `4.x.x` including latest `4.3.3`, I face this issue with `mouseover` and `mouseout `events in the following cases. 

1) I have a rectangle (graphic object) that moves back and forth through the canvas. When I place my cursor in the middle of the canvas, as the shape moves and crosses over my cursor, `mouseover` and `mouseout` events fail to register in a random pattern.

2) For the second case, the rectangle doesn't move, but this time I cross over the shape with my mouse. As I move my mouse fast ,  `mouseover `and `mouseout` events fail to register in a random pattern as well.

Here you can check the code https://jsbin.com/miwokog/edit?js,output

Below I also share javascript file

```js
/*
 * Mouse events randomly fail to register when delta is large or
 * rectangle width is small, or there's no animation but you
 * simply move your mouse fast over the rectangle.
 * Tested on versions 4.x.x including latest 4.3.3
 */

let delta = 10;
const width = 40;

const renderer = PIXI.autoDetectRenderer(400, 200)
document.body.appendChild(renderer.view)
const stage = new PIXI.Container()

const rect = new PIXI.Graphics()
rect.beginFill(0xf8d41f,1)
rect.drawRect(0, 0, width, 200)
rect.endFill()
rect.interactive = true
stage.addChild(rect)

rect.on('mouseover', function() {
  console.log("1 mouseover")
  this.alpha = 0.2
})
rect.on('mouseout', function() {
  console.log("2 mouseout")
  this.alpha = 1
})

animate()
function animate() {
  if (delta > 0 && rect.position.x >= 400) {
    delta = -delta
  }
  if (delta < 0 && rect.position.x <= 0) {
    delta = -delta
  }
  rect.x += delta
  renderer.render(stage)
  requestAnimationFrame(animate)
}
```

**Comment by staff0rd** [comment]
This is a limitation of javascript, not pixi.  The [event loop](https://developer.mozilla.org/en/docs/Web/JavaScript/EventLoop) means that new events can only be raised after current events have finished processing.

Check out [this pen](https://codepen.io/staff0rd/pen/NdjvmP?editors=0010) I've made to illustrate what's happening.  Open the console and move your mouse around over the two rabbits.  You'll see events being raised very quickly against the second rabbit, but very slowly against the first rabbit, even if you move the mouse slowly.

On inspection, the first rabbit has some code in its event handler that attempts to block the javascript thread from returning.  However, if that code wasn't there, then the handler would still only run as fast as whatever javascript engine you're using and whatever hardware you've got can run it.  Because you can't rely on what a user's browser/javascript engine/hardware is, it means you can't guarantee that an event will be raised for every input you throw at the dom.

In your case, if you must know that the mouse has very quickly moved from one side of the screen to the other (so fast that no event was raised), then you'll have to implement your own checks where you poll for the position of the mouse.

**Comment by serkandurusoy** [comment]
Hey @staff0rd I'm interested in this, too and thank you for taking the time to provide an explanation.

One thing that puzzles me with this example is the fact that it is so simple and there's almost nothing (_almost_), to hold up the event loop before it gives pixi a chance to detect and fire the registered mouse events. That example above has problems even at very slow paces.

In a project I had to resort to handling mouse events on the canvas object, getting the mouse position and delivering it down to the object tree where mouse-sensitive objects can know if they're being interacted with.

Surprizingly, that turned out to be much more performant, with little to no false negatives. But the problem was, the wiring required to make that happen, painted me in some other corner.

Tl:dr;
Mouse events handled on the canvas object and passed down as mouse positions perform better than individual handlers on pixi objects. Would this be expected behavior?

**Comment by staff0rd** [comment]
@serkandurusoy That would not be my expectation considering that pixi binds directly to events raised by the canvas element.  See `addEvents` in the [InteractionManager](https://github.com/pixijs/pixi.js/blob/dev/src/interaction/InteractionManager.js). 

Are you able to maybe supply a [pen](https://codepen.io) shows this performance discrepancy against the latest [dev build](//d157l7jdn8e5sf.cloudfront.net/dev/pixi.js)?

**Comment by serkandurusoy** [comment]
Hey @staff0rd thanks for the pointer. (And no I have not tried the dev build, is there anything new in there interactionwise that might change this for me?)

I am now suspecting that https://github.com/pixijs/pixi.js/blob/dev/src/interaction/InteractionManager.js#L809 is what may be causing the discrepancy because I have a graph where any number of nodes and/or children may or may not be interactive based on some ux context which gets passed down from a state manager.

I also happen to have simultaneous wheel/scroll handlers which need to sync up.

In the meantime, as wheel/scroll happens, or by external reactive data changes, the scene graph can get updated and the new objects may have different interactivity properties.

It is kind of hard to translate all that into a small pen that would demonstrate an isolated case, since I believe the case is probably cultivated by the combination of all of the above.

And perhaps it is already a better idea to handle the events externally.

I'm currently at odds about whether or not I want to use canvas/webgl or maybe I should refactor into dom and divs. Depending on the decision, I'll try to report back here with any new findings I may have.
